### PR TITLE
LPD-86525 search-experiences-service: Use path=null for empty reverse_nested aggregation config

### PR DESCRIPTION
**Pull request by jasmyt**
Ticket: https://liferay.atlassian.net/browse/LPD-86525

This fixes the case where an empty config for a reverse_nested aggregation in a search blueprint resolves to an empty string for the path attribute, which is the incorrect input to Elasticsearch and breaks the aggregation behavior.

### CODE REVIEW

File: modules/dxp/apps/search-experiences/search-experiences-service/src/main/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverter.java [hunk 1]
```diff
@@ -1177,7 +1177,8 @@ private RangeAggregation _toRangeAggregation(
 	private ReverseNestedAggregation _toReverseNestedAggregation(
 		JSONObject jsonObject, String name) {
 
-		return _aggregations.reverseNested(name, jsonObject.getString("path"));
+		return _aggregations.reverseNested(
+			name, jsonObject.getString("path", null));
```
[the remark this answers is not in the document]
Comment (jasmyt) [reply modules/dxp/apps/search-experiences/search-experiences-service/src/main/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverter.java:1181]: not sure that's necessary

**Comment by jasmyt** [comment]
ci:test:sf

File: modules/dxp/apps/search-experiences/search-experiences-service/src/main/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverter.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (joshuacords) [reply modules/dxp/apps/search-experiences/search-experiences-service/src/main/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverter.java:1181]: From the code, it does look like if the path is set to `""` that it would still be broken. But the case being fixed is that no path is set, but jsonObject.getString("path") defaults to empty string. I think this is fine as well because I don't think we should cover someone purposefully setting path to `""`.

**Comment by joshuacords** [comment]
The code looks good and you can send a hotfix, but we do want some type of test - we probably need a AggregationWrapperConverterTest unit test.

**Comment by jasmyt** [comment]
ci:test:sf

**Comment by jasmyt** [comment]
ci:test:sf

File: modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+public class AggregationWrapperConverterTest {
+
+	@ClassRule
+	@Rule
+	public static final LiferayUnitTestRule liferayUnitTestRule =
+		LiferayUnitTestRule.INSTANCE;
+
+	@Before
+	public void setUp() {
+		_aggregationWrapperConverter = new AggregationWrapperConverter(
```
Comment (rodrigoguedes) [inline modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java]: Since we don't need to maintain the state of the `_aggregationWrapperConverter`, I think we can instantiate it directly in a final attribute, like this:

```
	private final AggregationWrapperConverter _aggregationWrapperConverter =
		new AggregationWrapperConverter(
			new AggregationsImpl(), Mockito.mock(HighlightConverter.class),
			Mockito.mock(QueryConverter.class),
			Mockito.mock(ScriptConverter.class),
			Mockito.mock(SignificanceHeuristics.class),
			Mockito.mock(Sorts.class));
```

File: modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+			new AggregationsImpl(), Mockito.mock(HighlightConverter.class),
+			Mockito.mock(QueryConverter.class),
+			Mockito.mock(ScriptConverter.class),
+			Mockito.mock(SignificanceHeuristics.class),
+			Mockito.mock(Sorts.class));
+	}
+
+	@Test
+	public void testReverseNestedAggregationConfiguration() {
+		String aggregationType = "reverse_nested";
```
Comment (rodrigoguedes) [inline modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java]: I think we can create a constant to avoid duplication.

File: modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+		ReverseNestedAggregation reverseNestedAggregation =
+			(ReverseNestedAggregation)aggregation;
+
+		Assert.assertNull(reverseNestedAggregation.getPath());
+	}
+
+	private Aggregation _getAggregation(
+		String aggregationType, JSONObject aggConfigJSONObject) {
+
+		AggregationWrapper wrapper =
```
Comment (rodrigoguedes) [inline modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java]: change from `wrapper` to `aggregationWrapper`.

Use variables without abbreviations; I've seen PRs rejected for using abbreviations.

File: modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+		Assert.assertTrue(aggregation instanceof ReverseNestedAggregation);
+
+		ReverseNestedAggregation reverseNestedAggregation =
+			(ReverseNestedAggregation)aggregation;
+
+		Assert.assertNull(reverseNestedAggregation.getPath());
+	}
+
+	private Aggregation _getAggregation(
```
Comment (rodrigoguedes) [inline modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java]: We can remove the `aggregationType` parameter and directly use the constant suggested here: https://github.com/liferay-search/liferay-portal/pull/1874/changes#r3125982179

File: modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+		String path = RandomTestUtil.randomString();
+
+		JSONObject aggConfigJSONObject = JSONUtil.put(
+			aggregationType, JSONUtil.put("path", path));
+
+		Aggregation aggregation = _getAggregation(
+			aggregationType, aggConfigJSONObject);
+
+		Assert.assertTrue(aggregation instanceof ReverseNestedAggregation);
```
Comment (rodrigoguedes) [inline modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java]: This works, but I believe receiving a ClassCastException would provide more details about the problem, and the line below already does that. Does that make sense?

This change could also make our test shorter, for example:
```
		String path = RandomTestUtil.randomString();

		ReverseNestedAggregation reverseNestedAggregation =
			(ReverseNestedAggregation)_getAggregation(
				JSONUtil.put("path", path));

		Assert.assertEquals(path, reverseNestedAggregation.getPath());
```

**Review by rodrigoguedes — CHANGES_REQUESTED** [review]

**Comment by jasmyt** [comment]
ci:test:sf

File: modules/dxp/apps/search-experiences/search-experiences-api/src/main/java/com/liferay/search/experiences/constants/SXPConstants.java [hunk 7]
```diff
@@ -12,4 +12,7 @@ public class SXPConstants {
 
 	public static final String RESOURCE_NAME = "com.liferay.search.experiences";
 
+	public static final String REVERSE_NESTED_AGGREGATION_TYPE =
```
Comment (rodrigoguedes) [inline modules/dxp/apps/search-experiences/search-experiences-api/src/main/java/com/liferay/search/experiences/constants/SXPConstants.java]: This constant is only being used for testing, so it doesn't make sense for it to be here.
There are two options:

a) Instead of using the literal value that is still in AggregationWrapperConverter.java:142, change to using this constant REVERSE_NESTED_AGGREGATION_TYPE.

b) Remove this constant from SXPConstants and follow the pattern used by other types "literal usage" and leave the public API minimal without changes.

My opinion is that we don't need this constant in SXPConstants, my suggestion is to use option b).

Rethinking this, I don't think we even need this constant in the test; I'll explain the following suggestions better.

File: modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+public class AggregationWrapperConverterTest {
+
+	@ClassRule
+	@Rule
+	public static final LiferayUnitTestRule liferayUnitTestRule =
+		LiferayUnitTestRule.INSTANCE;
+
+	@Test
+	public void testReverseNestedAggregationConfiguration() {
+		String path = RandomTestUtil.randomString();
```
Comment (rodrigoguedes) [inline modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java]: My opinion in the previous review was to leave as little code as possible.

```
String path = RandomTestUtil.randomString();

ReverseNestedAggregation reverseNestedAggregation =
	_getReverseNestedAggregation(JSONUtil.put("path", path));

Assert.assertEquals(path, reverseNestedAggregation.getPath());
```

This way we have the following improvements:

* We concentrate the cast in a single point, within _getReverseNestedAggregation.
* We don't need to pass the type because the _getReverseNestedAggregation method already knows which one to use; the name itself indicates this.
* We don't need to have an additional aggregation variable.

File: modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+ */
+public class AggregationWrapperConverterTest {
+
+	@ClassRule
+	@Rule
+	public static final LiferayUnitTestRule liferayUnitTestRule =
+		LiferayUnitTestRule.INSTANCE;
+
+	@Test
+	public void testReverseNestedAggregationConfiguration() {
```
Comment (rodrigoguedes) [inline modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java]: The suffix is ​​too generic; we could have a more descriptive name for the behavior of the test.
We could change `testReverseNestedAggregationConfiguration` to `testReverseNestedAggregationWithPath`.

File: modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+				JSONUtil.put("path", path)));
+
+		ReverseNestedAggregation reverseNestedAggregation =
+			(ReverseNestedAggregation)aggregation;
+
+		Assert.assertEquals(path, reverseNestedAggregation.getPath());
+	}
+
+	@Test
+	public void testReverseNestedAggregationConfigurationEmpty() {
```
Comment (rodrigoguedes) [inline modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java]: We could change `testReverseNestedAggregationConfigurationEmpty` to `testReverseNestedAggregationWithoutPath`.

File: modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+		Aggregation aggregation = _getReverseNestedAggregation(
+			JSONUtil.put(
+				SXPConstants.REVERSE_NESTED_AGGREGATION_TYPE,
+				new JSONObjectImpl()));
+
+		ReverseNestedAggregation reverseNestedAggregation =
+			(ReverseNestedAggregation)aggregation;
+
+		Assert.assertNull(reverseNestedAggregation.getPath());
```
Comment (rodrigoguedes) [inline modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java]: The same intention as the previous method is to leave as little code as possible.

```
ReverseNestedAggregation reverseNestedAggregation =
	_getReverseNestedAggregation(JSONFactoryUtil.createJSONObject());

Assert.assertNull(reverseNestedAggregation.getPath());
```

One additional point of improvement in this method is: instead of using the `new JSONObjectImpl()` implementation, we can use `JSONFactoryUtil.createJSONObject()`, which is part of the public API and makes the test independent of the implementation. It is also the most commonly used Liferay pattern in tests when an empty value is needed.

File: modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+		AggregationWrapper aggregationWrapper =
+			_aggregationWrapperConverter.toAggregationWrapper(
+				aggConfigJSONObject.getJSONObject(
+					SXPConstants.REVERSE_NESTED_AGGREGATION_TYPE),
+				RandomTestUtil.randomString(),
+				SXPConstants.REVERSE_NESTED_AGGREGATION_TYPE);
+
+		return aggregationWrapper.getAggregation();
+	}
```
Comment (rodrigoguedes) [inline modules/dxp/apps/search-experiences/search-experiences-service/src/test/java/com/liferay/search/experiences/internal/blueprint/aggregation/AggregationWrapperConverterTest.java]: I suggest reusing things:

```
	private ReverseNestedAggregation _getReverseNestedAggregation(
		JSONObject jsonObject) {

		AggregationWrapper aggregationWrapper =
			_aggregationWrapperConverter.toAggregationWrapper(
				jsonObject, RandomTestUtil.randomString(), "reverse_nested");

		return (ReverseNestedAggregation)aggregationWrapper.getAggregation();
	}
```

Here we focus on two things:
* the "reverse_nested" type (in this case, we don't need a constant because there's no duplication)
* Casting of the ReverseNestedAggregation type

**Review by rodrigoguedes — CHANGES_REQUESTED** [review]

**Comment by jasmyt** [comment]
ci:test:sf

**Comment by jasmyt** [comment]
thanks for all the detailed feedback @rodrigoguedes

**Review by rodrigoguedes — APPROVED** [review]
LGTM

**Comment by jasmyt** [comment]
ci:forward
